rapids-cmake/cpm/nvtx3.cmake

```diff
@@ -0,0 +1,107 @@
+#=============================================================================
+# Copyright (c) 2024, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#=============================================================================
+include_guard(GLOBAL)
+
+#[=======================================================================[.rst:
+rapids_cpm_nvtx3
+----------------
+
+.. versionadded:: v24.06.00
+
+Allow projects to find `NVTX3` via `CPM` with built-in tracking of dependencies
+for correct export support.
+
+Uses the version of nvtx3 :ref:`specified in the version file <cpm_versions>` for consistency
+across all RAPIDS projects.
+
+.. code-block:: cmake
+
+  rapids_cpm_nvtx3( [BUILD_EXPORT_SET <export-name>]
+                    [INSTALL_EXPORT_SET <export-name>]
+                    [<CPM_ARGS> ...])
+
+.. |PKG_NAME| replace:: nvtx3
+.. include:: common_package_args.txt
+
+Result Targets
+^^^^^^^^^^^^^^
+  nvtx3-c, nvtx3-cpp targets will be created
+
+Result Variables
+^^^^^^^^^^^^^^^^
+  :cmake:variable:`NVTX3_SOURCE_DIR` is set to the path to the source directory of nvtx3.
+  :cmake:variable:`NVTX3_BINARY_DIR` is set to the path to the build directory of nvtx3.
+  :cmake:variable:`NVTX3_ADDED`      is set to a true value if nvtx3 has not been added before.
+  :cmake:variable:`NVTX3_VERSION`    is set to the version of nvtx3 specified by the versions.json.
+
+#]=======================================================================]
+function(rapids_cpm_nvtx3)
+  list(APPEND CMAKE_MESSAGE_CONTEXT "rapids.cpm.nvtx3")
+
+  set(options)
+  set(one_value USE_PROPRIETARY_BINARY BUILD_EXPORT_SET INSTALL_EXPORT_SET)
+  set(multi_value)
+  cmake_parse_arguments(_RAPIDS "${options}" "${one_value}" "${multi_value}" ${ARGN})
+
+  # Fix up _RAPIDS_UNPARSED_ARGUMENTS to have EXPORT_SETS as this is need for rapids_cpm_find
+  if(_RAPIDS_INSTALL_EXPORT_SET)
+    list(APPEND _RAPIDS_EXPORT_ARGUMENTS INSTALL_EXPORT_SET ${_RAPIDS_INSTALL_EXPORT_SET})
+  endif()
+  if(_RAPIDS_BUILD_EXPORT_SET)
+    list(APPEND _RAPIDS_EXPORT_ARGUMENTS BUILD_EXPORT_SET ${_RAPIDS_BUILD_EXPORT_SET})
+  endif()
+  set(_RAPIDS_UNPARSED_ARGUMENTS ${_RAPIDS_EXPORT_ARGUMENTS})
+
+  set(to_install OFF)
+  if(INSTALL_EXPORT_SET IN_LIST ARGN)
+    set(to_install ON)
+  endif()
```

Comment:
We can get rid of this block if we just condition the block below on `_RAPIDS_INSTALL_EXPORT_SET`.

Reply:
See other thread: https://github.com/rapidsai/rapids-cmake/pull/606/files#r1598804744

Reply:
this whole block needs to be moved to be after `rapids_cpm_package_details` and updated to also check the state of `exclude`. See cuco.cmake for a reference